server/src/main/java/org/opensearch/common/logging/DeprecatedMessage.java

```diff
@@ -77,6 +82,34 @@ private static Map<String, Object> fieldMap(String key, String xOpaqueId) {
     }
 
     public boolean isAlreadyLogged() {
-        return !keys.add(keyWithXOpaqueId);
+        return !keyDedupeCache.add(keyWithXOpaqueId);
+    }
+
+    private static class KeyDedupeCache {
+        private static final Object VALUE = new Object();
+        private final Cache<Integer, Object> cache;
+
+        private KeyDedupeCache() {
+            cache = CacheBuilder.<Integer, Object>builder()
+                .weigher((s, o) -> 1) // Entries are a fixed size (integers)
+                .setMaximumWeight(MAX_DEDUPE_CACHE_ENTRIES)
```

Comment:
Should we add in a time-based cache expire policy as well?

Reply:
I'm leaning towards simplicity here and l don't think we need it. Also if you enable time-based eviction then every `get()` call will have to invoke `System.nanoTime()` adding more overhead. Do you think it is necessary?